File: java/mil/nga/giat/mage/sdk/fetch/LocationServerFetch.java
package mil.nga.giat.mage.sdk.fetch;

import android.content.Context;
import android.util.Log;

import java.util.Collection;
import java.util.Date;

import mil.nga.giat.mage.sdk.datastore.location.Location;
import mil.nga.giat.mage.sdk.datastore.location.LocationHelper;
import mil.nga.giat.mage.sdk.datastore.location.LocationProperty;
import mil.nga.giat.mage.sdk.datastore.user.Event;
import mil.nga.giat.mage.sdk.datastore.user.EventHelper;
import mil.nga.giat.mage.sdk.datastore.user.User;
import mil.nga.giat.mage.sdk.datastore.user.UserHelper;
import mil.nga.giat.mage.sdk.exceptions.UserException;
import mil.nga.giat.mage.sdk.http.resource.LocationResource;

public class LocationServerFetch extends AbstractServerFetch {

	private static final String LOG_NAME = LocationServerFetch.class.getName();

	private UserHelper userHelper;
	private UserServerFetch userFetch;
	private LocationHelper locationHelper;
	private LocationResource locationResource;

	public LocationServerFetch(Context context) {
		super(context);

		userHelper = UserHelper.getInstance(context);
		locationHelper = LocationHelper.getInstance(context);
		locationResource = new LocationResource(context);
		userFetch = new UserServerFetch(mContext);
	}

	public void fetch() {
		User currentUser = null;
		try {
			currentUser = userHelper.readCurrentUser();
		} catch (UserException e) {
			Log.e(LOG_NAME, "Error rading current user.", e);
		}

		Event event = EventHelper.getInstance(mContext).getCurrentEvent();
		Collection<Location> locations = locationResource.getLocations(event);
		for (Location location : locations) {

			// make sure that the user exists and is persisted in the local data-store
			String userId = null;
			LocationProperty userIdProperty = location.getPropertiesMap().get("userId");
			if (userIdProperty != null) {
				userId = userIdProperty.getValue().toString();
			}
			try {
				if (userId != null) {
					User user = userHelper.read(userId);
					// TODO : test the timer to make sure users are updated as needed!
					final long sixHoursInMilliseconds = 6 * 60 * 60 * 1000;
					if (user == null || (new Date()).after(new Date(user.getFetchedDate().getTime() + sixHoursInMilliseconds))) {
						// get any users that were not recognized or expired
						userFetch.fetch(userId);
						user = userHelper.read(userId);
					}
					location.setUser(user);

					// if there is no existing location, create one
					Location l = locationHelper.read(location.getRemoteId());
					if (l == null) {
						// delete old location and create new one
						if (user != null) {
							// don't pull your own locations
							if (!user.equals(currentUser)) {
								userId = String.valueOf(user.getId());
								location = locationHelper.create(location);
								locationHelper.deleteUserLocations(userId, true, location.getEvent());
							}
						} else {
							Log.w(LOG_NAME, "A location with no user was found and discarded.  User id: " + userId);
						}
					}
				}
			} catch (Exception e) {
				Log.e(LOG_NAME, "There was a failure while performing an Location Fetch operation.", e);
			}
		}
	}
}

File: mage/src/main/java/mil/nga/giat/mage/location/LocationServerFetch.kt
package mil.nga.giat.mage.location

import android.content.Context
import android.util.Log
import mil.nga.giat.mage.sdk.datastore.location.LocationHelper
import mil.nga.giat.mage.sdk.datastore.user.EventHelper
import mil.nga.giat.mage.sdk.datastore.user.User
import mil.nga.giat.mage.sdk.datastore.user.UserHelper
import mil.nga.giat.mage.sdk.exceptions.UserException
import mil.nga.giat.mage.sdk.fetch.UserServerFetch
import mil.nga.giat.mage.sdk.http.resource.LocationResource
import java.util.*

class LocationServerFetch(val context: Context) {

    companion object {
        private val LOG_NAME = mil.nga.giat.mage.location.LocationServerFetch::class.java.name
    }

    private val userHelper: UserHelper = UserHelper.getInstance(context)
    private val userFetch: UserServerFetch = UserServerFetch(context)
    private val locationHelper: LocationHelper = LocationHelper.getInstance(context)
    private val locationResource: LocationResource = LocationResource(context)

    fun fetch() {
        var currentUser: User? = null
        try {
            currentUser = userHelper.readCurrentUser()
        } catch (e: UserException) {
            Log.e(LOG_NAME, "Error reading current user.", e)
        }

        val event = EventHelper.getInstance(context).currentEvent
        try {
            val locations = locationResource.getLocations(event)
            for (location in locations) {
                // make sure that the user exists and is persisted in the local data-store
                var userId: String? = null
                val userIdProperty = location.propertiesMap["userId"]
                if (userIdProperty != null) {
                    userId = userIdProperty.value.toString()
                }
                if (userId != null) {
                    var user: User? = userHelper.read(userId)
                    // TODO : test the timer to make sure users are updated as needed!
                    val sixHoursInMilliseconds = (6 * 60 * 60 * 1000).toLong()
                    if (user == null || Date().after(Date(user.fetchedDate.time + sixHoursInMilliseconds))) {
                        // get any users that were not recognized or expired
                        Log.d(LOG_NAME, "User for location is null or stale, re-pulling")
                        userFetch.fetch(userId)
                        user = userHelper.read(userId)
                    }
                    location.user = user

                    // if there is no existing location, create one
                    val l = locationHelper.read(location.remoteId)
                    if (l == null) {
                        // delete old location and create new one
                        if (user != null) {
                            // don't pull your own locations
                            if (user != currentUser) {
                                userId = user.id.toString()
                                val newLocation = locationHelper.create(location)
                                locationHelper.deleteUserLocations(userId, true, newLocation.event)
                            }
                        } else {
                            Log.w(LOG_NAME, "A location with no user was found and discarded.  User id: $userId")
                        }
                    }
                }
            }
        } catch(e: Exception) {
            Log.e(LOG_NAME, "Failed to fetch user locations from server", e)
        }

    }
}
